src/utils/auth.ts

```diff
@@ -0,0 +1,225 @@
+import {
+  AuthenticationResult,
+  BrowserAuthError,
+  BrowserAuthErrorMessage,
+  Configuration,
+  PublicClientApplication,
+  RedirectRequest,
+  SilentRequest,
+  SsoSilentRequest,
+} from '@azure/msal-browser';
+import { AccountInfo, InteractionRequiredAuthError } from '@azure/msal-browser';
+import {
+  loginRequest,
+  tokenRequest,
+  silentRequest,
+  msalConfig,
+} from '../config/authConfig';
+
+class AuthProvider {
+  private publicClient: PublicClientApplication;
+
+  private _account: AccountInfo | null = null;
+  private _token: AuthenticationResult | null = null;
+
+  private loginPromise: Promise<void> | null = null;
+  private initPromise: Promise<void[]>;
+
+  get isAuthenticated() {
+    return this._token && this._token.expiresOn > new Date();
+  }
+
+  get account() {
+    return this._account;
+  }
+
+  get token() {
+    return this._token;
+  }
+
+  constructor(config: Configuration) {
+    this.publicClient = new PublicClientApplication(config);
+
+    console.log('starting auth provider');
+
+    const initPromises: Promise<void>[] = [];
+
+    initPromises.push(
+      (async (publicClient: PublicClientApplication) => {
+        try {
+          const response = await publicClient.handleRedirectPromise();
+          this.handleLoginResponse(response, null);
+        } catch (err) {
+          this.handleLoginResponse(null, err);
+        }
+      })(this.publicClient)
+    );
+
+    const account = this.getActiveAccount();
+
+    if (account) {
+      console.log('found cached account', account);
+      this._account = account;
+    }
+
+    initPromises.push(
+      (async () => {
+        if (this.loginPromise) await this.loginPromise;
+        this.loginPromise = this.login(loginRequest);
+        await this.loginPromise;
+      })()
+    );
+
+    this.initPromise = Promise.all(initPromises);
+  }
+
+  private handleLoginResponse(
+    response: AuthenticationResult | null,
+    err: Error | null
+  ) {
+    if (err) {
+      console.error('login error', err);
+      return;
+    }
+
+    console.log('login result', response);
+
+    const account = this.getActiveAccount();
+
+    if (response) {
+      console.log('login success, has response', account);
+      this._account = account;
+      this._token = response;
+    } else {
```

Comment:
I mentioned yesterday that it's actually possible to `throw null`, in which case we could end up here. Anyway this is based on documented/sample code, and I see it happening locally: 
![image](https://user-images.githubusercontent.com/5425986/101750092-d53ab300-3ace-11eb-9764-8a7c4b60a8d1.png)

Before merge I will be documenting the flow in comments so that we have clear expectations of how things work, since auth flows like this can be complex. Usually end up discovering some inconsistencies that way, so maybe I'll find out the reason later, but like I said all this logic is based on samples and MSAL docs, so usually there is some reason, and it may or may not be applicable to us. But it works 👍